File: 0022_generate_parentheses/main.go
package main

func findCombinations(left, right int, res *[]string, comb string) {
	if left == 0 && right == 0 && len(comb) != 0 {
		*res = append(*res, comb)
		return
	}
	if left > 0 {
		findCombinations(left-1, right, res, comb+"(")
	}
	if right > 0 && right > left {
		findCombinations(left, right-1, res, comb+")")
	}
}

func generateParenthesis(n int) []string {
	res := []string{}
	findCombinations(n, n, &res, "")
	return res
}

func main() {

}

File: 0022_generate_parentheses/main.py
class Solution:
    res = []

    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        self.res = []
        self.findCombinations(n, n, "")
        return self.res
    
    def findCombinations(self, left, right, comb):
        if left == 0 and right == 0 and len(comb) != 0 {
            self.res.append(comb)
            return
        }
        if left > 0:
            self.findCombinations(left-1, right, comb+"(")
        if right > 0 and right > left:
            self.findCombinations(left, right-1, comb+")")